examples/silentpayments.c

```diff
@@ -0,0 +1,417 @@
+        for (i = 0; i < N_TX_OUTPUTS; i++) {
+            printf("    ");
+            printf("%s : ", address_amounts[i]);
+            secp256k1_xonly_pubkey_serialize(ctx, xonly_print, &tx_outputs[i]);
+            print_hex(xonly_print, sizeof(xonly_print));
+        }
+        /* It's best practice to try to clear secrets from memory after using them.
+         * This is done because some bugs can allow an attacker to leak memory, for
+         * example through "out of bounds" array access (see Heartbleed), Or the OS
+         * swapping them to disk. Hence, we overwrite the secret key buffer with zeros.
+         *
+         * Here we are preventing these writes from being optimized out, as any good compiler
+         * will remove any writes that aren't used. */
+        for (i = 0; i < N_TX_INPUTS; i++) {
+            secure_erase(&sender_seckeys[i], sizeof(sender_seckeys[i]));
+        }
+    }
+
+    /*** Receiving ***/
+    {
+        /*** Transaction data ***
+         *
+         * Here we create a few global variables to represent the transaction data:
+         *
+         *     1. The transaction inputs, `tx_input_ptrs`
+         *     2. The transaction outputs, `tx_output_ptrs`
+         *
+         * These will be used to demonstrate scanning as a full node and scanning as a light client.
+         */
+        const secp256k1_xonly_pubkey *tx_input_ptrs[N_TX_INPUTS];
+        const secp256k1_xonly_pubkey *tx_output_ptrs[N_TX_OUTPUTS];
+        unsigned char light_client_data33[33];
+
+        for (i = 0; i < N_TX_INPUTS; i++) {
+            tx_input_ptrs[i] = &tx_inputs[i];
+        }
+        for (i = 0; i < N_TX_OUTPUTS; i++) {
+            tx_output_ptrs[i] = &tx_outputs[i];
+        }
+
+        /*** Scanning with labels as a full node (Bob) ***
+         *
+         * Since Bob has access to the full transaction, scanning is simple:
+         *
+         *     1. Collect the relevant data from the transaction inputs and call
+         *        `secp256k1_silentpayments_recipient_public_data_create`
+         *     2. Call `secp256k1_silentpayments_recipient_scan_outputs`
+         *
+         * Additionally, since Bob has access to the full transaction outputs when scanning its easy for him
+         * to scan with labels, as demonstrated below. For efficient scanning, Bob keeps a cache of
+         * every label he has previously used and uses a callback to check if a potential label exists
+         * in his cache. Since the labels are created using an incremental integer `m`, if Bob ever
+         * forgets how many labels he has previously used, he can pregenerate a large number of
+         * labels (e.g. 0..100_000) and use that while scanning.
+         */
+        {
+            secp256k1_silentpayments_found_output found_outputs[N_TX_OUTPUTS];
+            secp256k1_silentpayments_found_output *found_output_ptrs[N_TX_OUTPUTS];
+            secp256k1_silentpayments_public_data public_data;
```

Comment:
Can be demystified with something like this:

```c
/** Public data is an x-only (?) public key that can be generated
  * for each transaction without knowledge of either the input
  * private keys or the recipient silent payment address. This
  * function performs ECDH on it and checks if any found_output_ptrs
  * contains the resulting x-only pubkey - indication it's a silent
  * payment to us.
  * See also light client scanning below.
  */
```

Reply:
It might even make sense to move the light client example up, since by necessity it explains some of the things that happen automagically.

Reply:
> ```c
> /** Public data is an x-only (?) public key that can be generated
>   * for each transaction without knowledge of either the input
>   * private keys or the recipient silent payment address. This
>   * function performs ECDH on it and checks if any found_output_ptrs
>   * contains the resulting x-only pubkey - indication it's a silent
>   * payment to us.
>   * See also light client scanning below.
>   */
> ```

Demystifying is a good idea, but I think we shouldn't mix API usage with protocol internals. (If people want to know how it works under the hood, they should just read the BIP.)

 - The public data is public data that can be generated for each transaction without knowledge of either the input private keys or the recipient silent payment address. Whether it's a public key is not relevant for API users. 
 - This checks if it's (potential) silent payment to us. The fact that it does an ECDH key exchange is a detail of the protocol.

Reply:
I wrote most of this based on my own initial confusing with the example code, and I have read the BIP. Agree it shouldn't duplicate the whole thing, just enough so people understand what the code is trying to do.

Reply:
I agree its better to refer people to the BIP if they are interested in the internals. For `public_data` specifically, I think the documentation for `_public_data_create` should be sufficient for explaining to the caller what a `public_data` object is.

As @Sjors pointed out in a different comment, `smallest_outpoint` isn't well documented, but aside from that I think referring to this function should be sufficient for the caller.

Reply:
The term "public data" doesn't occur in the BIP. Even if you point to the section "Creating outputs" it's not obvious. But it might be better to clarify this in the BIP instead of in the demo.

Reply:
Updated the wording and added documentation for `smallest_outpoint`.